### ISSUE TITLE
[functionality]Alternate:  The Game Loop!

### ISSUE THREAD

**Issue by Sillinde** [Sillinde/Sillinde.github.io#9]
Got a suggestion from someone, and gonna try an alternate means to an end with the game loop.  I guess using intervals like that are REAL bad for games trying to operate in the background, so working off the system clock, instead, seems to be the way to go.

var newTime;
var oldTime = new Date();
var systemCycle = 0;
// Update every second
var tick = 1000;
function processSystemCycle() {
    newTime = new Date();
    systemCycle += newTime.getTime() - oldTime.getTime();
    if (systemCycle >= tick){
        var numUpdates = systemCycle % tick;
        // Update everything numUpdates times
        systemCycle %= tick;
    }
    oldTime = new Date();
}

Thanks to grammarxcore for the code in /r/incremental_games FBF thread.  I think I know where these chunks go, so I'll play with that later.

What this will do is, instead of running from the setInterval bit, and counting out ms, it will run the check whenever it can, see how much time has passed since the last update, and, if it has been over 1 second, update by however much it needs to in order to catch up.  I'm not sure if this will also allow "offline" progress, but I'd rather not have that, so, we just keep all the time variables out of the save function.  I mean, if we save it, as well, then when it gets loaded, it will see that "OMG THREE DAYS HAVE PASSED" and explode.  Well, not literally, but, you know what I mean.

So, whenever we update resources, we multiply the resource gains (or deficits) by numUpdates.  For the most part, numUpdates should be "1".  Unless the player has been tabbed out, and the game is trying to run in the background.  Then, when they make it the active tab again, it'll see that numUpdates equals, let's say, 600 (they were surfing pr0n for 10 minutes, or something).  It'll multiply all resources by 600.

But!  Before I get to this, I have some cleaning to do around the shop.

**Comment by Sillinde** [comment]
It's actually pretty freaking neat.  The more we've been doing this, the more I can look at code (what I've done, what you've done, and even other people's code), and actually kind of understand what is trying to happen.  It's a lot more complex than QBasic ever got, but I think I'm catching on.  :)

**Comment by TheAntonverse** [comment]
Will this new loop style still allow things like the hidden house button show up immediately upon meeting the resource prerequisites, or will there be a delay? A note about this, though, is that the showing of the button was pretty much for test purposes to see if I could do it. I keep forgetting that the code I write is mostly for testing/learning, and the actual flow of the game could be different :palm_tree:

**Comment by Sillinde** [comment]
Aye.  It'll constantly be running that loop, I think.  It's only in the section between
if (systemCycle >= tick){
var numUpdates = systemCycle % tick;
// Update everything numUpdates times
systemCycle %= tick;
}
that, that will run once per second.  Otherwise, I guess framerate will be "as fast as it can."  :dango: 

I'll try that out in a branch.  I think I made one already.  Fuck.  I dunno.  XD  But, yeah.  Both of the previous game loops were all from setInterval, I believe.  Your's from JSFiddle, and mine from a "so you want to build a game" thread.  XD

**Comment by Sillinde** [comment]
Oh, and in regards to things being written for testing, EVERYTHING I write is a test.  Shit.  I expect every change I make to break everything, somehow.  heheheheh

**Comment by TheAntonverse** [comment]
I just don't want to get to a point where our testing suddenly requires major rewrites in order to become the actual game because we strayed too far from the game's actual design :P

Seems we're making good progress though! It's like we're going down a checklist... Can we make buttons? :white_check_mark: Can we display and increase values when we click the buttons? :white_check_mark: Can we style the buttons so they're not awful looking default buttons? :white_check_mark: Can we make a functioning game loop? :white_check_mark: Can we hide/un-hide buttons based on conditions? :white_check_mark: 

Now we're still working on the loop, which is good. We're improving the performance as we go! I think once you've got that implemented, we need to sit down and draw out how we want the game screen to look. There's a lot of talk about separating buttons into relevant tabs. We'll also probably want an area to display game information as it happens (An adventuring party goes off to explore the nearby woods. / An adventuring party returns to town beaten up. / etcetc)

**Comment by Sillinde** [comment]
Yeah. Right now, the biggest reason to change the game loop is to allow it
to run in the background.  Not a huge issue, though. Like you say, is
rather not have to rewrite major chunks of code all the time. We currently
have a timing cycle working, so I've just been puttering with this just to
see if I can. Xd
